In [43]:
# Importing the necessary libraries for analysis
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import OneHotEncoder
from scipy.sparse import hstack

In [44]:
#Reading all the datasets
air_conditioners_df = pd.read_csv('Air Conditioners.csv')
all_cars_bikes_df = pd.read_csv('All Car and Motorbike Products.csv')
all_electronics_df = pd.read_csv('All Electronics.csv')
all_exercise_fitness_df = pd.read_csv('All Exercise and Fitness.csv')
all_grocery_df = pd.read_csv('All Grocery and Gourmet Foods.csv')
all_kitchen_df = pd.read_csv('All Home and Kitchen.csv')
all_pet_supplies_df = pd.read_csv('All Pet Supplies.csv')
all_sports_df = pd.read_csv('All Sports Fitness and Outdoors.csv')
all_amazon_df = pd.read_csv('Amazon-Products.csv')
all_fashion_df = pd.read_csv('Amazon Fashion.csv')
baby_bath_grooming_df = pd.read_csv('Baby Bath Skin and Grooming.csv')
baby_fashion_df = pd.read_csv('Baby Fashion.csv')
baby_products_df = pd.read_csv('Baby Products.csv')
backpacks_df = pd.read_csv('Backpacks.csv')
badminton_df = pd.read_csv('Badminton.csv')
bags_and_luggage_df = pd.read_csv('Bags and Luggage.csv')
ballerinas_df = pd.read_csv('Ballerinas.csv')
beauty_and_grooming = pd.read_csv('Beauty and Grooming.csv')
bedroom_linen_df = pd.read_csv('Bedroom Linen.csv')
camera_accessories_df = pd.read_csv('Camera Accessories.csv')
camera_df = pd.read_csv('Cameras.csv')
camping_and_hiking_df = pd.read_csv('Camping and Hiking.csv')
car_accessories_df = pd.read_csv('Car Accessories.csv')
car_and_bike_care_df = pd.read_csv('Car and Bike Care.csv')
car_electronics_df = pd.read_csv('Car Electronics.csv')
car_parts_df = pd.read_csv('Car Parts.csv')
cardio_equipment_df = pd.read_csv('Cardio Equipment.csv')
casual_shoes_df = pd.read_csv('Casual Shoes.csv')
clothing_df = pd.read_csv('Clothing.csv')
coffee_tea_beverages_df = pd.read_csv('Coffee Tea and Beverages.csv')
cricket_df = pd.read_csv('Cricket.csv')
cycling_df = pd.read_csv('Cycling.csv')
diapers_df = pd.read_csv('Diapers.csv')
diet_and_nutrition_df = pd.read_csv('Diet and Nutrition.csv')
dog_supplies_df = pd.read_csv('Dog supplies.csv')
ethnic_wear_df = pd.read_csv('Ethnic Wear.csv')
fashion_and_jewellery_df = pd.read_csv('Fashion and Silver Jewellery.csv')
fashion_sandals_df = pd.read_csv('Fashion Sandals.csv')
fitness_accessories_df = pd.read_csv('Fitness Accessories.csv')
football_df = pd.read_csv('Football.csv')
formal_shoes_df = pd.read_csv('Formal Shoes.csv')
furniture_df = pd.read_csv('Furniture.csv')
garden_and_outdoors_df = pd.read_csv('Garden and Outdoors.csv')
gold_and_diamond_jewellery_df = pd.read_csv('Gold and Diamond Jewellery.csv')
handbags_and_clutches_df = pd.read_csv('Handbags and Clutches.csv')
headphones_df = pd.read_csv('Headphones.csv')
health_and_personal_care_df = pd.read_csv('Health and Personal Care.csv')
heating_and_cooling_appliances_df = pd.read_csv('Heating and Cooling Appliances.csv')
home_audio_and_theater_df = pd.read_csv('Home Audio and Theater.csv')
home_decor_df = pd.read_csv('Home Dcor.csv')
home_entertainment_systems_df = pd.read_csv('Home Entertainment Systems.csv')
home_furnishing_df = pd.read_csv('Home Furnishing.csv')
home_improvement_df = pd.read_csv('Home Improvement.csv')
home_storage_df = pd.read_csv('Home Storage.csv')
household_supplies_df = pd.read_csv('Household Supplies.csv')
indoor_lighting_df = pd.read_csv('Indoor Lighting.csv')
industrial_and_scientific_supplies_df = pd.read_csv('Industrial and Scientific Supplies.csv')
innerwear_df = pd.read_csv('Innerwear.csv')
janitorial_and_sanitation_supplies_df = pd.read_csv('Janitorial and Sanitation Supplies.csv')
jeans_df = pd.read_csv('Jeans.csv')
jewellery_df = pd.read_csv('Jewellery.csv')
kids_clothing_df = pd.read_csv('Kids Clothing.csv')
kids_fashion_df = pd.read_csv('Kids Fashion.csv')
kids_shoes_df = pd.read_csv('Kids Shoes.csv')
kids_watches_df = pd.read_csv('Kids Watches.csv')
kitchen_and_dining_df = pd.read_csv('Kitchen and Dining.csv')
kitchen_and_storage_containers_df = pd.read_csv('Kitchen Storage and Containers.csv')
kitchen_and_home_appliances_df = pd.read_csv('Kitchen and Home Appliances.csv')
lab_and_scientific_df = pd.read_csv('Lab and Scientific.csv')
lingerie_and_nightwear_df = pd.read_csv('Lingerie and Nightwear.csv')
luxury_beauty_df = pd.read_csv('Luxury Beauty.csv')
makeup_df = pd.read_csv('Make-up.csv')
mens_fashion_df = pd.read_csv('Mens Fashion.csv')
motorbike_accessories_and_parts_df = pd.read_csv('Motorbike Accessories and Parts.csv')
musical_instruments_and_professional_audio_df = pd.read_csv('Musical Instruments and Professional Audio.csv')
nursing_and_feeding_df = pd.read_csv('Nursing and Feeding.csv')
personal_care_appliances_df = pd.read_csv('Personal Care Appliances.csv')
refrigerators_df = pd.read_csv('Refrigerators.csv')
rucksack_df = pd.read_csv('Rucksacks.csv')
running_df = pd.read_csv('Running.csv')
school_bags_df = pd.read_csv('School Bags.csv')
security_cameras_df = pd.read_csv('Security Cameras.csv')
sewing_and_craft_supplies = pd.read_csv('Sewing and Craft Supplies.csv')
shirts_df = pd.read_csv('Shirts.csv')
shoes_df = pd.read_csv('Shoes.csv')
snack_foods_df = pd.read_csv('Snack Foods.csv')
speakers_df = pd.read_csv('Speakers.csv')
sports_shoes_df = pd.read_csv('Sports Shoes.csv')
sportswear_df = pd.read_csv('Sportswear.csv')
strength_training_df = pd.read_csv('Strength Training.csv')
strollers_and_prams_df = pd.read_csv('Strollers and Prams.csv')
suitcases_and_trolley_bags_df = pd.read_csv('Suitcases and Trolley Bags.csv')
sunglasses_df = pd.read_csv('Sunglasses.csv')
tshirts_and_polos_df = pd.read_csv('T-shirts and Polos.csv')
televisions_df = pd.read_csv('Televisions.csv')
test_measure_and_inspect_df = pd.read_csv('Test Measure and Inspect.csv')
the_designer_boutique_df = pd.read_csv('The Designer Boutique.csv')
toys_and_games_df = pd.read_csv('Toys and Games.csv')
travel_accessories_df = pd.read_csv('Travel Accessories.csv')
travel_duffles_df = pd.read_csv('Travel Duffles.csv')
value_bazaar_df = pd.read_csv('Value Bazaar.csv')
wallets_df = pd.read_csv('Wallets.csv')
washing_machines_df = pd.read_csv('Washing Machines.csv')
watches_df = pd.read_csv('Watches.csv')
western_wear_df = pd.read_csv('Western Wear.csv')
womens_fashion_df = pd.read_csv('Womens Fashion.csv')
yoga_df = pd.read_csv('Yoga.csv')

In [45]:
# Combine the DataFrames vertically
df = pd.concat([
air_conditioners_df,
all_cars_bikes_df,
all_electronics_df,
all_exercise_fitness_df,
all_grocery_df,
all_kitchen_df,
all_pet_supplies_df,
all_sports_df,
all_amazon_df,
all_fashion_df,
baby_bath_grooming_df,
baby_fashion_df, 
baby_products_df, 
backpacks_df, 
badminton_df,
bags_and_luggage_df,
ballerinas_df, 
beauty_and_grooming,
bedroom_linen_df,
camera_accessories_df,
camera_df, 
camping_and_hiking_df,
car_accessories_df,
car_and_bike_care_df, 
car_electronics_df,
car_parts_df,
cardio_equipment_df,
casual_shoes_df,
clothing_df, 
coffee_tea_beverages_df,
cricket_df,
cycling_df,
diapers_df,
diet_and_nutrition_df,
dog_supplies_df,
ethnic_wear_df,
fashion_and_jewellery_df,
fashion_sandals_df,
fitness_accessories_df,
football_df,
formal_shoes_df,
furniture_df,
garden_and_outdoors_df,
gold_and_diamond_jewellery_df, 
handbags_and_clutches_df,
headphones_df,
health_and_personal_care_df, 
heating_and_cooling_appliances_df, 
home_audio_and_theater_df,
home_decor_df,
home_entertainment_systems_df,
home_furnishing_df,
home_improvement_df,
home_storage_df, 
household_supplies_df,
indoor_lighting_df,
industrial_and_scientific_supplies_df,
innerwear_df, 
janitorial_and_sanitation_supplies_df,
jeans_df,
jewellery_df,
kids_clothing_df, 
kids_fashion_df, 
kids_shoes_df,
kids_watches_df, 
kitchen_and_dining_df, 
kitchen_and_storage_containers_df, 
kitchen_and_home_appliances_df,
lab_and_scientific_df,
lingerie_and_nightwear_df,
luxury_beauty_df,
makeup_df,
mens_fashion_df,
motorbike_accessories_and_parts_df, 
musical_instruments_and_professional_audio_df, 
nursing_and_feeding_df,
personal_care_appliances_df,
refrigerators_df, 
rucksack_df,
running_df, 
school_bags_df, 
security_cameras_df,
sewing_and_craft_supplies,
shirts_df,
shoes_df, 
snack_foods_df, 
speakers_df,
sports_shoes_df,
sportswear_df,
strength_training_df,
strollers_and_prams_df,
suitcases_and_trolley_bags_df,
sunglasses_df, 
tshirts_and_polos_df,
televisions_df,
test_measure_and_inspect_df, 
the_designer_boutique_df,
toys_and_games_df, 
travel_accessories_df, 
travel_duffles_df, 
value_bazaar_df, 
wallets_df,
washing_machines_df,
watches_df,
western_wear_df, 
womens_fashion_df,
yoga_df,
], ignore_index=True)

# Display the shape of the combined DataFrame and the first few rows to confirm the combination
(df.shape, df.head())


((1093430, 10),
                                                 name main_category  \
 0  Lloyd 1.5 Ton 3 Star Inverter Split Ac (5 In 1...    appliances   
 1  LG 1.5 Ton 5 Star AI DUAL Inverter Split AC (C...    appliances   
 2  LG 1 Ton 4 Star Ai Dual Inverter Split Ac (Cop...    appliances   
 3  LG 1.5 Ton 3 Star AI DUAL Inverter Split AC (C...    appliances   
 4  Carrier 1.5 Ton 3 Star Inverter Split AC (Copp...    appliances   
 
        sub_category                                              image  \
 0  Air Conditioners  https://m.media-amazon.com/images/I/31UISB90sY...   
 1  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
 2  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
 3  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
 4  Air Conditioners  https://m.media-amazon.com/images/I/41lrtqXPiW...   
 
                                                 link ratings no_of_ratings  \
 0  https://www.amazon.in

In [46]:
df.head(5)

name main_category  \
0  Lloyd 1.5 Ton 3 Star Inverter Split Ac (5 In 1...    appliances   
1  LG 1.5 Ton 5 Star AI DUAL Inverter Split AC (C...    appliances   
2  LG 1 Ton 4 Star Ai Dual Inverter Split Ac (Cop...    appliances   
3  LG 1.5 Ton 3 Star AI DUAL Inverter Split AC (C...    appliances   
4  Carrier 1.5 Ton 3 Star Inverter Split AC (Copp...    appliances   

       sub_category                                              image  \
0  Air Conditioners  https://m.media-amazon.com/images/I/31UISB90sY...   
1  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
2  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
3  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
4  Air Conditioners  https://m.media-amazon.com/images/I/41lrtqXPiW...   

                                                link ratings no_of_ratings  \
0  https://www.amazon.in/Lloyd-Inverter-Convertib...     4.2         2,255   
1  https://www.amazon.in/LG-Convertible-Anti-Viru...     4.2         2,948   
2  https://www.amazon.in/LG-Inverter-Convertible-...     4.2         1,206   
3  https://www.amazon.in/LG-Convertible-Anti-Viru...     4.0            69   
4  https://www.amazon.in/Carrier-Inverter-Split-C...     4.1           630   

  discount_price actual_price  Unnamed: 0  
0        ₹32,999      ₹58,990         NaN  
1        ₹46,490      ₹75,990         NaN  
2        ₹34,490      ₹61,990         NaN  
3        ₹37,990      ₹68,990         NaN  
4        ₹34,490      ₹67,790         NaN

In [47]:
Col_to_drop = df.drop(['image', 'link','Unnamed: 0'],axis=1, inplace=True)

In [48]:
df.shape

(1093430, 7)

In [49]:
df.describe()
df.dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1093430 entries, 0 to 1093429
Data columns (total 7 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   name            1093430 non-null  object
 1   main_category   1093430 non-null  object
 2   sub_category    1093430 non-null  object
 3   ratings         742330 non-null   object
 4   no_of_ratings   742330 non-null   object
 5   discount_price  971507 non-null   object
 6   actual_price    1057905 non-null  object
dtypes: object(7)
memory usage: 58.4+ MB


In [50]:
#Changing the data type of discounted price and actual price

df['discount_price'] = df['discount_price'].str.replace("₹",'')
df['discount_price'] = df['discount_price'].str.replace(",",'')
df['discount_price'] = df['discount_price'].astype('float64')

df['actual_price'] = df['actual_price'].str.replace("₹",'')
df['actual_price'] = df['actual_price'].str.replace(",",'')
df['actual_price'] = df['actual_price'].astype('float64')

In [51]:
df.dtypes

name               object
main_category      object
sub_category       object
ratings            object
no_of_ratings      object
discount_price    float64
actual_price      float64
dtype: object

In [52]:
df['no_of_ratings'] = df['no_of_ratings'].str.replace(r'[^\d.]', '', regex=True)
df['no_of_ratings'] = pd.to_numeric(df['no_of_ratings'], errors='coerce')
df['ratings'] = pd.to_numeric(df['ratings'], errors='coerce')

In [53]:
df.dtypes

name               object
main_category      object
sub_category       object
ratings           float64
no_of_ratings     float64
discount_price    float64
actual_price      float64
dtype: object

In [54]:
df.head()

name main_category  \
0  Lloyd 1.5 Ton 3 Star Inverter Split Ac (5 In 1...    appliances   
1  LG 1.5 Ton 5 Star AI DUAL Inverter Split AC (C...    appliances   
2  LG 1 Ton 4 Star Ai Dual Inverter Split Ac (Cop...    appliances   
3  LG 1.5 Ton 3 Star AI DUAL Inverter Split AC (C...    appliances   
4  Carrier 1.5 Ton 3 Star Inverter Split AC (Copp...    appliances   

       sub_category  ratings  no_of_ratings  discount_price  actual_price  
0  Air Conditioners      4.2         2255.0         32999.0       58990.0  
1  Air Conditioners      4.2         2948.0         46490.0       75990.0  
2  Air Conditioners      4.2         1206.0         34490.0       61990.0  
3  Air Conditioners      4.0           69.0         37990.0       68990.0  
4  Air Conditioners      4.1          630.0         34490.0       67790.0

In [55]:
df.dropna(subset=['discount_price'], inplace = True)
df.dropna(subset=['ratings'], inplace = True)
df.dropna(subset=['actual_price'], inplace = True)

In [56]:
missing_values = df[['ratings', 'no_of_ratings', 'discount_price', 'actual_price']].isnull().sum()
missing_values

ratings           0
no_of_ratings     0
discount_price    0
actual_price      0
dtype: int64

In [58]:
df.shape

(661112, 7)

In [59]:
df.to_csv('cleaned.csv', index=False)

In [37]:
# Feature engineering (e.g., create a discount rate feature)
df['discount_rate'] = (df['actual_price'] - df['discount_price']) / df['actual_price']

# Text processing for product names
tfidf = TfidfVectorizer(max_features=100)
tfidf_matrix = tfidf.fit_transform(df['name'])

# One-hot encoding for categorical variables
ohe = OneHotEncoder()
encoded_categories = ohe.fit_transform(df[['main_category', 'sub_category']])

# Combine all features
X = hstack([tfidf_matrix, encoded_categories, df[['actual_price', 'discount_rate']]])

# Split the df into training and testing sets
y = df['discount_price']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train a Random Forest model
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Make predictions and evaluate the model
predictions = rf_model.predict(X_test)
mse = mean_squared_error(y_test, predictions)
rmse = mse ** 0.5

print(f'RMSE: {rmse}')


RMSE: 378.2479860854766


An RMSE (Root Mean Squared Error) value of 378.2479860854766 in the context of your dataset's pricing prediction model gives us an indication of the model's prediction accuracy in terms of how close the predicted prices are to the actual prices, on average. Here's what it specifically tells us and how you might approach finding the accuracy of the predictions:

Interpretation of RMSE: 378.2479860854766
Average Prediction Error: The RMSE value represents the standard deviation of the residuals (prediction errors). In this case, an RMSE of 378.25 suggests that, on average, the model's predictions are off by about 378.25 units of your dataset's currency from the actual discount prices.

Contextual Significance: The significance of an RMSE value depends on the context, including the scale and range of your target variable (e.g., discount prices). If the prices in your dataset typically range in the thousands, an RMSE of 378.25 could be considered relatively small and indicative of a reasonably accurate model. However, if your prices are generally lower, this RMSE might signify a less accurate model.

Model Evaluation: RMSE is a commonly used metric for evaluating the performance of regression models. A lower RMSE value indicates better model performance. However, it's important to compare this value against a baseline model or other models to contextualize its performance.

Sensitivity to Outliers: RMSE is sensitive to outliers because it squares the errors. Therefore, large errors due to outliers can disproportionately affect the RMSE value. If your dataset contains outliers in the target variable, they could significantly impact the RMSE.

In [42]:
from sklearn.metrics import r2_score
r2 = r2_score(y_test, predictions)
r2

0.9961868015827496

Accuracy of the Prediction
In regression tasks, "accuracy" as it's typically understood in classification tasks
(i.e., the proportion of correct predictions) doesn't directly apply. However, there are several ways to assess the performance and predictive accuracy of regression models:

R-squared (Coefficient of Determination): 
This metric provides a measure of how well the observed outcomes are replicated by the model, based on the proportion of total variation of outcomes explained by the model. An R-squared value closer to 1 indicates a model that accurately predicts the target variable.

In [43]:
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test, predictions)
mae

5.267640530769988